src/Microsoft.AspNetCore.Razor.Language/Legacy/MarkupParser.cs

```diff
@@ -0,0 +1,916 @@
+
+                    if (seenEndTag)
+                    {
+                        var endTag = ParseEndTag();
+                        builder.Add(endTag);
+                    }
+                }
+            }
+            else
+            {
+                // Parsing an end tag.
+                var endTag = ParseEndTag();
+                builder.Add(endTag);
+            }
+        }
+
+        private MarkupTagBlockSyntax ParseStartTag(out bool expectsBody)
+        {
+            expectsBody = true;
+            using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+            {
+                var tagBuilder = pooledResult.Builder;
+
+                AcceptAndMoveNext(); // Accept '<'
+                ParseOptionalBangEscape(tagBuilder);
+
+                if (At(SyntaxKind.Text))
+                {
+                    var tagName = CurrentToken.Content;
+                    if (_voidElements.Contains(tagName))
+                    {
+                        // This is a void element.
+                        expectsBody = false;
+                    }
+                }
+                TryAccept(SyntaxKind.Text);
+
+                // Parse the contents of a tag like attributes.
+                ParseTagContent(tagBuilder);
+
+                if (TryAccept(SyntaxKind.ForwardSlash))
+                {
+                    // This is a self closing tag.
+                    expectsBody = false;
+                }
+                TryAccept(SyntaxKind.CloseAngle);
+
+                // End tag block
+                tagBuilder.Add(OutputAsMarkupLiteral());
+                var tagBlock = SyntaxFactory.MarkupTagBlock(tagBuilder.ToList());
+                return tagBlock;
+            }
+        }
+
+        private MarkupTagBlockSyntax ParseEndTag()
+        {
+            // This section can accept things like: '</p  >' or '</p>' etc.
+            Assert(SyntaxKind.OpenAngle);
+
+            using (var pooledResult = Pool.Allocate<RazorSyntaxNode>())
+            {
+                var tagBuilder = pooledResult.Builder;
+
+                AcceptAndMoveNext(); // Accept '<'
+                TryAccept(SyntaxKind.ForwardSlash);
+
+                // Whitespace here is invalid (according to the spec)
+                ParseOptionalBangEscape(tagBuilder);
+                TryAccept(SyntaxKind.Text);
+                TryAccept(SyntaxKind.Whitespace);
+                TryAccept(SyntaxKind.CloseAngle);
+
+                // End tag block
+                tagBuilder.Add(OutputAsMarkupLiteral());
+                var tagBlock = SyntaxFactory.MarkupTagBlock(tagBuilder.ToList());
+                return tagBlock;
+            }
+        }
+
+        private SyntaxList<RazorSyntaxNode> ParseTagBody(string tagName, out bool seenEndTag)
+        {
+            // No-op here for now.
+            seenEndTag = false;
+            return null;
+        }
+
+        private string GetTagName(MarkupTagBlockSyntax tagBlock)
```

Comment:
Should this be a method on `MarkupTagBlockSyntax`?

Reply:
https://github.com/aspnet/Razor/pull/2750/files#r236458761